### PR TITLE
feature(http-client): add basic http-client implementation

### PR DESCRIPTION
**Pull request by matheus-consoli**
The first implementation of a mini HTTP client based on `reqwest`'s API

The client uses hyper, rustls, and async-std.

The client supports bytes and `json` (as an opt-in feature).

Depends on #215

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "fluvio-future"
-version = "0.6.1"
+version = "0.6.2"
```
Comment (sehz) [inline Cargo.toml:3]: since 0.6.1 is not released, no version change is needed

File: src/http_client/client.rs [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+pub struct Client {
+    pub(crate) hyper: HyperClient,
+}
+
+impl Client {
+    pub fn new() -> Self {
+        ClientBuilder::default().build()
+    }
+
+    pub fn get(&self, uri: impl AsRef<str>) -> RequestBuilder {
```
Comment (sehz) [inline src/http_client/client.rs:23]: is there reason for another layout of abstraction?  Maybe this shouldn't be called `Client`,  Something like `ReadOnlyClient` or specific purpose

File: src/http_client/request.rs [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+        <HeaderValue as TryFrom<V>>::Error: Into<http::Error>,
+    {
+        self.req_builder = self.req_builder.header(key, value);
+        self
+    }
+
+    pub async fn send(self) -> Result<Response<Body>, Error> {
+        let req = self
+            .req_builder
+            .header(http::header::USER_AGENT, "fluvio-mini-http/0.1")
```
Comment (sehz) [inline src/http_client/request.rs:37]: is this just for simple get?  then let's just move to read only

File: tests/http_client.rs [hunk 4]
```diff
@@ -0,0 +1,57 @@
+#![cfg(all(any(unix, windows), feature = "http-client-json"))]
+
+use fluvio_future::http_client::{self, ResponseExt, StatusCode};
+
```
Comment (sehz) [inline tests/http_client.rs:4]: integrated test should not have outside dependency which could be offline or behavior changed.  Please take look at other service testing.  This means crate fake http server so we can control exact behavior to test

**Review by sehz — CHANGES_REQUESTED** [review]

File: tests/http_client.rs (hunk 4 above)
Comment (digikata) [reply tests/http_client.rs:4]: There is a nice reference unit test in the hub-pkg-server repo https://github.com/infinyon/hub-service/blob/4841b25d4a13c6f2f035d2310c2600c3d1341b8c/crates/hub-pkg-server/src/util.rs#L314

Comment (digikata) [reply tests/http_client.rs:4]: Setting up a standalone https server with a root-certificate derived cert is time consuming, can we mark ignore on external sites referenced in this test, and keep the https://infinyon domain sites in for this test.

File: src/http_client/client.rs (hunk 2 above)
Comment (matheus-consoli) [reply src/http_client/client.rs]: I removed the `ClientBuilder` layer and moved its contents to `Client` itself

File: src/http_client/request.rs (hunk 3 above)
Comment (matheus-consoli) [reply src/http_client/request.rs:37]: I'm not sure what you meant by "read only" here

But I prefer to keep the `RequestBuilder` the way it is:
- it's closer to the `reqwest` API
- it's more extensible if we want to add new features
- it has better usability (it's not possible to `Client::get().get().get()...`, for example)

File: src/http_client/client.rs [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    }
+}
+
+impl Client {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    pub fn get(&self, uri: impl AsRef<str>) -> RequestBuilder {
+        let req = http::request::Builder::new().uri(Uri::from_str(uri.as_ref()).unwrap());
```
Comment (sehz) [inline src/http_client/client.rs:56]: remove unwrap

File: src/http_client/request.rs [hunk 6]
```diff
@@ -0,0 +1,83 @@
+use std::{fmt, future::Future, pin::Pin};
+
+use anyhow::{anyhow, Error};
+use futures_util::TryFutureExt;
+use http::{request::Builder, HeaderName, HeaderValue};
+use hyper::{body::Bytes, Body, Response};
+
+use super::client::Client;
+
+pub struct RequestBuilder {
```
Comment (sehz) [inline src/http_client/request.rs:10]: use https://crates.io/crates/derive_builder,  it's been used in the `fluvio` successfully

File: src/http_client/request.rs (hunk 3 above)
Comment (sehz) [reply src/http_client/request.rs:37]: k

File: src/http_client/request.rs [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+        HeaderValue: TryFrom<V>,
+        <HeaderValue as TryFrom<V>>::Error: Into<http::Error>,
+    {
+        self.req_builder = self.req_builder.header(key, value);
+        self
+    }
+
+    pub async fn send(self) -> Result<Response<Body>, Error> {
+        let req = self
+            .req_builder
```
Comment (sehz) [inline src/http_client/request.rs:36]: let's not set default here.  Put all defaults in the builder so it's configurable.  using `derive_builder` will simplify effort

**Review by sehz — CHANGES_REQUESTED** [review]

**Comment by digikata** [comment]
Continued on https://github.com/infinyon/future-aio/pull/219

**Pull request closed without merging**